src/server/game/Spells/Spell.cpp

```diff
@@ -2448,7 +2448,7 @@ void Spell::DoAllEffectOnTarget(TargetInfo* target)
         // Needs to be called after dealing damage/healing to not remove breaking on damage auras
         DoTriggersOnSpellHit(spellHitTarget, mask);
 
-        // if target is fallged for pvp also flag caster if a player
+        // if target is flagged for pvp also flag caster if a player
         if (unit->IsPvP() && m_caster->GetTypeId() == TYPEID_PLAYER)
             m_caster->ToPlayer()->UpdatePvP(true);
```

Comment:
I'm on Linux using Geany/make at the moment, so the first step for Visual Studio for me would be wine shenanigans or getting a Windows installation.

I found the core reason for the problem. I was so focused on Spell Hit that I forgot the original bug: Mind Control not flagging the caster for PvP.

`spellHitTarget` is working exactly as intended. The problem is that the following important things are happening:

1. Player casts Mind Control, eventually leading to `DoAllEffectOnTarget`.
2. `DoSpellHitOnUnit is called, and my PvP flag update copied from the bottom of `DoAllEffectOnTarget` takes effect.
3. The various effects of Mind Control are applied. This is where the problem is. After the target is charmed, the victim becomes an **unflagged** minion.
4. Exit back out to `DoAllEffectOnTarget`.
5. Continue to the PvP check block. If the spell successfully hit (note that misses and immunes still work as intended), then the target is not flagged and the method continues without flipping the caster over.

So this looks like it's much more specific of a problem than I first thought.

To resolve this, would there be any disadvantage have the original PvP check set a boolean flag in `DoAllEffectOnTarget` immediately before the call to `DoSpellHitOnUnit`, then act on it where it currently is?

```cpp
if (unit->IsPvP() && m_caster->GetTypeId() == TYPEID_PLAYER)
    enablePvP = true;
...
if (enablePvP)
   m_caster->ToPlayer()->UpdatePvP(true);
```